org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/views/IssueLocationsView.java

```diff
@@ -0,0 +1,402 @@
+        continue;
+      }
+      // Cast needed for older Eclipse versions
+      IFile file = (IFile) editorInput.getAdapter(IFile.class);
+      if (file != null && sonarlintMarker.getResource().equals(file)) {
+        IEditorPart editorPart = editor.getEditor(false);
+        if (editorPart instanceof ITextEditor) {
+          this.currentEditor = (ITextEditor) editorPart;
+          this.currentDoc = currentEditor.getDocumentProvider().getDocument(editorPart.getEditorInput());
+          this.currentFile = file;
+          locationsViewer.setInput(sonarlintMarker);
+          try {
+            ShowIssueFlowsMarkerResolver.toggleAnnotations((IMarker) locationsViewer.getInput(), currentEditor, true);
+          } catch (BadPositionCategoryException e) {
+            SonarLintLogger.get().error("Unable to show annotations", e);
+          }
+          return;
+        }
+      }
+    }
+    showMarkerWithNoEditor(sonarlintMarker);
+  }
+
+  private void showMarkerWithNoEditor(IMarker sonarlintMarker) {
+    clearCurrentEditor();
+    locationsViewer.setInput(sonarlintMarker);
+  }
+
+  private void clearInput() {
+    clearCurrentEditor();
+    locationsViewer.setInput(null);
+  }
+
+  private void clearCurrentEditor() {
+    this.currentEditor = null;
+    this.currentDoc = null;
+    this.currentFile = null;
+  }
+
+  @Override
+  public void selectionChanged(IWorkbenchPart part, ISelection selection) {
+    IMarker selectedMarker = AbstractSonarWebView.findSelectedSonarIssue(selection);
+    if (selectedMarker != null && selectedMarker != locationsViewer.getInput()) {
+      setInput(selectedMarker);
+    }
+  }
+
+  private void startListeningForSelectionChanges() {
+    getSite().getPage().addPostSelectionListener(this);
+  }
+
+  private void stopListeningForSelectionChanges() {
+    getSite().getPage().removePostSelectionListener(this);
+  }
+
+  @Override
+  public void createPartControl(Composite parent) {
+    createToolbar();
+    Tree tree = new Tree(parent, SWT.MULTI);
+    locationsViewer = new LocationsViewer(tree);
+    locationsViewer.setContentProvider(new LocationsProvider());
+    locationsViewer.setLabelProvider(new LocationsTreeLabelProvider());
+    locationsViewer.addSelectionChangedListener(event -> {
+      ISelection selection = event.getSelection();
+      if (selection instanceof IStructuredSelection) {
+        Object firstElement = ((IStructuredSelection) selection).getFirstElement();
+        if (firstElement == null) {
+          return;
+        }
+        IWorkbenchPage page = PlatformUI.getWorkbench().getActiveWorkbenchWindow().getActivePage();
+        try {
+          if (firstElement instanceof RootNode) {
+            IDE.openEditor(page, ((RootNode) firstElement).getRoot());
+          } else if (firstElement instanceof ExtraPosition) {
+            IEditorPart editor = IDE.openEditor(page, currentFile);
+            if (editor instanceof ITextEditor) {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png 'Severity: CRITICAL') Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS134)